### PR TITLE
Support ref fields in `System.Reflection.Metadata.Ecma335.BlobEncoder`.

### LINKED ISSUE #68309: [API Proposal]: Support ref fields in System.Reflection.Metadata.Ecma335.BlobEncoder
### Background and motivation

Provide methods in `System.Reflection.Metadata.Ecma335.BlobEncoder` to more directly support emitting ref fields, consistent with the corresponding APIs provided for emitting ref parameters and ref locals.

### API Proposal

Provide a `FieldTypeEncoder` type similar to `ParameterTypeEncoder` and `LocalVariableTypeEncoder`.
```C#
namespace System.Reflection.Metadata.Ecma335
{
    public readonly struct BlobEncoder
    {
        public FieldTypeEncoder FieldTypeEncoder();
    }

    public readonly struct FieldTypeEncoder
    {
        public BlobBuilder Builder { get; }

        public FieldTypeEncoder(BlobBuilder builder);

        public CustomModifiersEncoder CustomModifiers();

        public SignatureTypeEncoder Type(bool isByRef = false);

        public void TypedReference();
    }
}
```

### API Usage

```C#
var blobBuilder = new BlobBuilder();
var encoder = new BlobEncoder(blobBuilder);

var fieldEncoder = encoder.FieldTypeEncoder();

// Ref custom modifiers: modopt(object)
var customModifiersEncoder = fieldEncoder.CustomModifiers();
customModifiersEncoder.AddModifier(MetadataTokens.TypeReferenceHandle(2), isOptional: true);

// Type: int&
var typeEncoder = fieldEncoder.Type(isByRef: true);
typeEncoder.PrimitiveType(PrimitiveTypeCode.Int32);
```

### Alternative Designs

The alternative is the existing API which requires emitting `SignatureTypeCode.ByReference` and creating a `CustomModifierEncoder` explicitly. The existing API is sufficient, but inconsistent with the support for ref parameters and ref locals. 

Using the existing API:
```C#
var blobBuilder = new BlobBuilder();
var encoder = new BlobEncoder(blobBuilder);

var typeEncoder = encoder.FieldSignature();

// Ref custom modifiers: modopt(object)
var customModifiersEncoder = new CustomModifiersEncoder(blobBuilder);
customModifiersEncoder.AddModifier(MetadataTokens.TypeReferenceHandle(2), isOptional: true);

// Type: int&
typeEncoder.Builder.WriteByte((byte)SignatureTypeCode.ByReference);
typeEncoder.PrimitiveType(PrimitiveTypeCode.Int32);
```

### Risks

`BlobEncoder` already includes a `FieldSignature` method that returns an encoder for a field type, so having two distinct field encoder methods may be confusing.
```C#
    public readonly struct BlobEncoder
    {
        public SignatureTypeEncoder FieldSignature();
        public FieldTypeEncoder FieldTypeEncoder();
    }
```

### PR DESCRIPTION
**Pull request by teo-tsirpanis**
Fixes #68309.
A test case was added.

### CODE REVIEW

**Comment by AaronRobinsonMSFT** [comment]
The ref assembly must also be updated.

File: src/libraries/System.Reflection.Metadata/tests/Metadata/Ecma335/Encoding/BlobEncodersTests.cs [hunk 1]
```diff
@@ -27,6 +27,21 @@ public void BlobEncoder_FieldSignature()
...
+            var e = new BlobEncoder();
+            Assert.Same(b, e.Builder);
+
+            var f = e.Field();
+            f.CustomModifiers()
+                .AddModifier(MetadataTokens.TypeDefinitionHandle(1), isOptional: true)
+                .AddModifier(MetadataTokens.TypeDefinitionHandle(2), isOptional: false);
+            f.Type(isByRef: true).Object();
+            AssertEx.Equal(new byte[] { 0x06, 0x20, 0x04, 0x1F, 0x08, 0x10, 0x1C }, b.ToArray());
+        }
```
Comment (cston) [inline src/libraries/System.Reflection.Metadata/tests/Metadata/Ecma335/Encoding/BlobEncodersTests.cs:43]: > } [](http://example.com/codeflow?start=7&length=2)

Please consider adding a test case calling `f.Type()`, with no explicit `isByRef` parameter. Feel free to add it to this test after calling `b.Clear()`, to match the style of other test methods.

**Comment by teo-tsirpanis** [comment]
Feedback was addressed.

**Review by cston — APPROVED** [review]
Thanks @teo-tsirpanis.

**Comment by AaronRobinsonMSFT** [comment]
This PR is in a weird state due to the Azure issues over the weekend. @teo-tsirpanis Can you push up a minor change just to re-trigger AzDO?

**Comment by teo-tsirpanis** [comment]
@AaronRobinsonMSFT I closed and reopened the PR, the CI got triggered and it is now green.

**Pull request merged**